### PR TITLE
feat: add header and footer

### PR DESCRIPTION
**Pull request by KenanHaciyev**

### CODE REVIEW

**Comment by Tsurhanau** [comment]
please name the PR according to:
https://github.com/JS-Machine-University/JSMU-WEB/wiki/Git-rules#merge-requests-params

File: projects/core/src/lib/components/header/header.component.html [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		</ul>
+
+		<div class="jsmu-header__menu-wrapper">
+			<menu class="jsmu-header__menu">
+				<li class="jsmu-header__menu-item list-item" *ngFor="let menuItem of menuItems">
+					<a
+						*ngIf="menuItem.activeOptions; else noOptions"
+						class="jsmu-header__menu-link link"
+						[routerLink]="menuItem.routerLink"
+						[routerLinkActiveOptions]="menuItem.activeOptions"
```
Comment (Tsurhanau) [inline projects/core/src/lib/components/header/header.component.html:22]: add here just [routerLinkActiveOptions]="{ exact: true }" 

and remove from activeOptions from menuItem

File: src/assets/objects/header.objects.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+export interface menuItemsInterface {
+	itemName: string;
+	routerLink: string;
+	activeOptions?: Object;
+}
+
+export interface Object {
+	exact: boolean;
+}
+
```
Comment (Tsurhanau) [inline src/assets/objects/header.objects.ts:10]: it's a redundant

File: src/assets/objects/header.objects.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+export const menuItems: menuItemsInterface[] = [
```
Comment (Tsurhanau) [inline src/assets/objects/header.objects.ts:11]: interfaces shouldn't be in assets, move it in models folder

File: projects/common-components/src/lib/button/button.component.scss [hunk 4]
```diff
@@ -77,3 +78,38 @@
     }
   }
 }
+
+:host.authorization {
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/common-components/src/lib/button/button.component.scss]: done

File: projects/core/src/lib/authorization/components/role-select/role-select.component.scss [hunk 5]
```diff
@@ -2,7 +2,7 @@
 @import "../../../../../../../src/assets/styles/mixins";
 
 .role-select {
-	@include wide-screen;
+	//@include wide-screen;
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/authorization/components/role-select/role-select.component.scss]: done

File: projects/core/src/lib/authorization/components/sign-out/sign-out.component.html [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<div class="jsmu-sign-out">
+	<jsmu-button (click)="logout()" class="sign-in__button authorization">Logout</jsmu-button>
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/authorization/components/sign-out/sign-out.component.html]: done

File: projects/core/src/lib/authorization/components/sign-out/sign-out.component.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { Component } from "@angular/core";
+import { AuthService } from "../../services/auth/auth.service";
+import { Observable } from "rxjs";
+
+@Component({
+	selector: "jsmu-sign-out",
+	templateUrl: "./sign-out.component.html",
+	styleUrls: ["./sign-out.component.scss"]
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/authorization/components/sign-out/sign-out.component.ts]: I changed everywhere to onPush except of one component -  it was some problems with this that is why i remained by default

File: projects/core/src/lib/authorization/components/sign-out/sign-out.component.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+})
+export class SignOutComponent {
+	public isLogged!: Observable<any>;
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/authorization/components/sign-out/sign-out.component.ts]: done

File: projects/core/src/lib/authorization/components/sign-out/sign-out.component.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+	constructor(private authService: AuthService) {}
+
+	public logout() {
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/authorization/components/sign-out/sign-out.component.ts:13]: done

File: projects/core/src/lib/components/footer/footer.component.html [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+				<li class="jsmu-footer__nav-item list-item">
+					<a class="jsmu-footer__nav-link link" href="">About</a>
+				</li>
+				<li class="jsmu-footer__nav-item list-item">
+					<a class="jsmu-footer__nav-link link" href="">Blog</a>
+				</li>
+				<li class="jsmu-footer__nav-item list-item">
+					<a class="jsmu-footer__nav-link link" href="">Contact</a>
+				</li>
+			</ul>
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/footer/footer.component.html]: done

File: projects/core/src/lib/components/footer/footer.component.html [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+					<a>
+						<i class="fa-brands fa-square-facebook"></i>
+					</a>
+				</li>
+			</ul>
+		</nav>
+	</div>
+
+	<div class="jsmu-footer__text">
+		<p>© 2023 EPAM Systems, Inc. All Rights Reserved.</p>
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/footer/footer.component.html]: done

File: projects/core/src/lib/components/footer/footer.component.scss [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	padding: 3px 15px;
+	font-size: 1.1vw;
+	border-top: 1px rgba(54, 55, 51, 0.8) solid;
+	text-align: center;
+	display: flex;
+	flex-direction: column;
+	justify-content: center;
+	align-items: center;
+
+	.link {
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/footer/footer.component.scss]: done

File: projects/core/src/lib/components/footer/footer.component.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import { Component } from "@angular/core";
+
+@Component({
+	selector: "jsmu-footer",
+	templateUrl: "./footer.component.html",
+	styleUrls: ["./footer.component.scss"]
+})
+export class FooterComponent {
+	date: Date = new Date();
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/footer/footer.component.ts]: done

File: projects/core/src/lib/components/header/header.component.html [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+			<li class="jsmu-header__avatar-item">
+				<a href="#" class="jsmu-header__avatar-link link">
+					<img
+						src="/assets/images/cat.jpg"
+						alt="profile"
+						class="jsmu-header__avatar-img"
+					/>
+				</a>
+			</li>
+		</ul>
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.html]: done

File: projects/core/src/lib/components/header/header.component.scss [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+  .list-item:last-child{
+    border-right: none;
+    padding-right: 0;
+  }
+
+  .link {
+    color: rgba(203,203,197,0.85);
+    text-decoration: none;
+  }
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.scss]: done

File: projects/core/src/lib/components/header/header.component.scss [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  &__login-item, &__logout-item {
+    list-style: none;
+  }
+
+  .active-link {
+    color: #fff;
+  }
+}
+
+
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.scss:111]: done

File: projects/core/src/lib/components/header/header.component.ts [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+import { Subject, takeUntil } from "rxjs";
+import { AuthService } from "../../authorization/services/auth/auth.service";
+import { UsersDataService } from "../../services/users.data.service";
+import { User } from "../../authorization/models/user";
+import { menuItems } from "../../../../../../src/assets/objects/header.objects";
+
+@Component({
+	selector: "jsmu-header",
+	templateUrl: "./header.component.html",
+	styleUrls: ["./header.component.scss"]
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.ts:11]: done

File: projects/core/src/lib/components/header/header.component.html (hunk 1 above)
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.html]: done

File: src/assets/objects/header.objects.ts (hunk 2 above)
Comment (KenanHaciyev) [reply src/assets/objects/header.objects.ts]: done

File: src/assets/objects/header.objects.ts (hunk 3 above)
Comment (KenanHaciyev) [reply src/assets/objects/header.objects.ts]: done

File: projects/core/src/lib/authorization/components/sign-out/sign-out.component.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Tsurhanau) [reply projects/core/src/lib/authorization/components/sign-out/sign-out.component.ts:13]: not

File: projects/core/src/lib/components/header/header.component.ts [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+	public log() {
+		this.authService
+			.getUser()
+			.pipe(takeUntil(this.destroy$))
+			.subscribe((sUser) => {
+				this.usersDataServ.getUserById(sUser.uid!).subscribe((user: User | null) => {
+					this.isLoggedIn = !!user;
+					return !!user;
+				});
+			});
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.ts]: done

File: projects/common-components/src/lib/avatar/avatar.component.scss [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+    @include width-height(100%, 100%);
+    object-fit: cover;
+    border-radius: 50%;
+    border: 2px lawngreen solid;
+    display: block;
+    margin: 0 auto;
+  }
+}
+
+@media (max-width: 992px) {
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/common-components/src/lib/avatar/avatar.component.scss]: done

File: projects/common-components/src/lib/avatar/avatar.component.ts [hunk 20]
```diff
@@ -0,0 +1,9 @@
+import { ChangeDetectionStrategy, Component } from "@angular/core";
+
+@Component({
+	selector: "jsmu-avatar",
+	templateUrl: "./avatar.component.html",
+	styleUrls: ["./avatar.component.scss"],
+	changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class AvatarComponent {}
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/common-components/src/lib/avatar/avatar.component.ts]: done

File: projects/common-components/src/lib/common-components.module.ts [hunk 21]
```diff
@@ -14,15 +20,27 @@ import { SortPipe } from "@jsmu/shared";
...
 		CommonComponentsComponent,
 		ButtonComponent,
 		TalksListComponent,
 		TalkItemComponent,
-		TalkItemsLineComponent
+		TalkItemsLineComponent,
+		NavigationItemComponent,
+		NavigationItemListComponent,
+		NavigationIconItemComponent,
+		NavigationIconItemListComponent,
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/common-components/src/lib/common-components.module.ts]: done

File: projects/core/src/lib/authorization/components/sign-in/sign-in.component.scss [hunk 22]
```diff
@@ -2,11 +2,12 @@
 @import "../../../../../../../src/assets/styles/mixins";
 
 .sign-in {
-	@include wide-screen;
+	//@include wide-screen;
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/authorization/components/sign-in/sign-in.component.scss]: done

File: projects/core/src/lib/components/header/header.component.html [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+				<ng-template #loggedOut>
+					<li class="jsmu-header__login-item">
+						<jsmu-sign-in></jsmu-sign-in>
+					</li>
+				</ng-template>
+			</ul>
+		</div>
+
+		<div
+			class="jsmu-header__burger-menu"
```
Comment (Tsurhanau) [inline projects/core/src/lib/components/header/header.component.html:36]: please use use the same naming burger-menu or sideBar

File: projects/core/src/lib/components/header/header.component.html [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+			[ngClass]="{active: sideBar}"
+			(click)="sideBar=!sideBar"
```
Comment (Tsurhanau) [inline projects/core/src/lib/components/header/header.component.html:38]: please move open logic to component and add unit test.
the approach with writing logic in the template is difficult to expand.

File: projects/core/src/lib/components/header/header.component.html [hunk 25]
```diff
@@ -0,0 +1,45 @@
+<header class="jsmu-header" [ngClass]="{mobile: sideBar}">
+	<nav class="jsmu-header__navigation">
```
Comment (Tsurhanau) [inline projects/core/src/lib/components/header/header.component.html:2]: why don't you reuse jsmu-navigation component? please fix

File: projects/core/src/lib/components/header/header.component.html [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+		<jsmu-avatar [avatarPath]="avatarPath"></jsmu-avatar>
```
Comment (Tsurhanau) [inline projects/core/src/lib/components/header/header.component.html:3]: why avatar is nested in nav? please fix

File: projects/core/src/lib/components/header/header.component.html [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+				<li class="jsmu-header__logout-item" *ngIf="isLoggedIn; else loggedOut">
+					<jsmu-sign-out></jsmu-sign-out>
+				</li>
+
+				<ng-template #loggedOut>
+					<li class="jsmu-header__login-item">
+						<jsmu-sign-in></jsmu-sign-in>
+					</li>
+				</ng-template>
+			</ul>
```
Comment (Tsurhanau) [inline projects/core/src/lib/components/header/header.component.html:32]: using ul > li for this block does not make sense

File: projects/core/src/lib/components/header/header.component.html [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+		</div>
+
+		<div
+			class="jsmu-header__burger-menu"
+			[ngClass]="{active: sideBar}"
+			(click)="sideBar=!sideBar"
+		>
+			<span class="jsmu-header__span"></span>
+			<span class="jsmu-header__span"></span>
+			<span class="jsmu-header__span"></span>
```
Comment (Tsurhanau) [inline projects/core/src/lib/components/header/header.component.html:42]: you can use structural directive for this for more convenient record.

File: projects/core/src/lib/components/header/header.component.scss [hunk 29]
```diff
@@ -0,0 +1,165 @@
...
+    list-style: none;
+    border-right: 3px rgba(6, 6, 18, 0.89) dotted;
+    padding-right: 1vw;
+
+    :hover {
+      color: $lightBlueColor;
+    }
+  }
+
+  &__list-item:last-child {
```
Comment (Tsurhanau) [inline projects/core/src/lib/components/header/header.component.scss:22]: please make nesting for :last-child  into  &__list-item

File: projects/core/src/lib/components/header/header.component.scss [hunk 30]
```diff
@@ -0,0 +1,165 @@
...
+      align-items: center;
+      position: absolute;
+      right: 10px;
+      top: 50%;
+      transform: translateY(-50%);
+      cursor: pointer;
+      @include width-height(25px, 28px);
+    }
+
+    &__span {
```
Comment (Tsurhanau) [inline projects/core/src/lib/components/header/header.component.scss:95]: please use a more appropriate class name

File: projects/common-components/src/lib/avatar/avatar.component.scss [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+			margin-top: 10px;
+			@include width-height(50px, 50px);
+		}
+	}
+
+	@media screen and (max-width: $widthSm) {
+		&__item {
+			margin-top: 3px;
+			@include width-height(40px, 40px);
+		}
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/common-components/src/lib/avatar/avatar.component.scss:29]: done

File: projects/common-components/src/lib/button/button.component.scss [hunk 32]
```diff
@@ -2,78 +2,87 @@
...
     }
+  }
 
-    &:hover:before {
-      @include width-height(35%, 65%);
-      @include top-left(57%, -14%);
+  @media (max-width: $widthSm) {
+    .button {
+      @include button-adaptivity(30px, 40px, 15px, 73%, -10%, 73%, -10%, 61%, -14%, 61%, -14%)
     }
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/common-components/src/lib/button/button.component.scss:80]: done

File: projects/core/src/lib/authorization/services/auth/auth.service.ts [hunk 33]
```diff
@@ -81,7 +80,7 @@ export class AuthService {
 	public signOut(): Promise<void> {
 		return this.afAuth.signOut().then(() => {
 			localStorage.removeItem("user");
-			this.router.navigate([Routes.SIGN_IN]);
+			this.router.navigate(["home-page"]);
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/authorization/services/auth/auth.service.ts:83]: done

File: projects/core/src/lib/components/footer/footer.component.scss [hunk 34]
```diff
@@ -0,0 +1,52 @@
...
+
+	&__date {
+		margin-bottom: 1vh;
+	}
+
+	&__content {
+		width: 30%;
+	}
+
+  .jsmu-footer__item-list {
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/footer/footer.component.scss:24]: done

File: projects/core/src/lib/components/header/header.component.html [hunk 35]
```diff
@@ -0,0 +1,45 @@
+<header class="jsmu-header" [ngClass]="{mobile: sideBar}">
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.html:1]: done

File: projects/core/src/lib/components/header/header.component.html (hunk 24 above)
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.html:38]: done

File: projects/core/src/lib/components/header/header.component.html (hunk 25 above)
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.html:2]: done

File: projects/core/src/lib/components/header/header.component.html (hunk 26 above)
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.html:3]: done

File: projects/core/src/lib/components/header/header.component.html (hunk 28 above)
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.html:42]: done

File: projects/core/src/lib/components/header/header.component.scss (hunk 29 above)
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.scss:22]: done

File: projects/common-components/src/lib/avatar/avatar.mixins.scss [hunk 36]
```diff
@@ -0,0 +1,8 @@
+@import "../../../../../src/assets/styles/mixins";
+
+@mixin margin-size($mt, $width, $height) {
+  &__item {
+    margin-top: $mt;
+    @include width-height($width, $height);
+  }
+}
```
Comment (Tsurhanau) [inline projects/common-components/src/lib/avatar/avatar.mixins.scss:8]: this mixin better define just in avatar.scss without separate file

File: projects/common-components/src/lib/navigation/navigation-item/navigation-item.component.ts [hunk 37]
```diff
@@ -0,0 +1,12 @@
...
+import { FooterItem } from "@jsmu/core";
+
+@Component({
+	selector: "jsmu-navigation-item",
+	templateUrl: "./navigation-item.component.html",
+	styleUrls: ["./navigation-item.component.scss"],
+	changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class NavigationItemComponent {
+	@Input() footerItem!: FooterItem;
```
Comment (Tsurhanau) [inline projects/common-components/src/lib/navigation/navigation-item/navigation-item.component.ts:11]: **footer**Item is not correct name, NavigationItem should not be tied to **footer**

File: projects/core/src/lib/components/header/header.component.ts [hunk 38]
```diff
@@ -0,0 +1,32 @@
...
+		this.isUserLogin();
+	}
+
+	public isUserLogin(): void {
+		this.usersDataServ.isUserLogin().subscribe((user) => {
+			this.isLoggedIn = !!user;
+		});
+	}
+
+	public toggleSideBar() {
```
Comment (Tsurhanau) [inline projects/core/src/lib/components/header/header.component.ts:29]: add return value :void for function

Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.ts]: done

File: projects/core/src/lib/models/footer.item.ts [hunk 39]
```diff
@@ -0,0 +1,3 @@
+export interface FooterItem {
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/models/footer.item.ts]: done

File: projects/core/src/lib/services/users.data.service.spec.ts [hunk 40]
```diff
@@ -6,17 +6,28 @@ import { HttpResponse } from "@angular/common/http";
...
 	let udb: UsersDataService;
 	let httpTestingController: HttpTestingController;
+	let fakeAuthServiceObj = jasmine.createSpyObj(["getUser"]);
+
 	const expectedUsers = [
 		{ uid: "0", email: "", name: "", photoURL: "", isVerified: false, role: {} },
 		{ uid: "1", email: "", name: "", photoURL: "", isVerified: false, role: {} }
 	] as User[];
 	beforeEach(() => {
+		// TestBed.overrideComponent(UsersDataService, {set: {providers: []}})
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/services/users.data.service.spec.ts]: done

File: projects/core/src/lib/authorization/guards/auth/auth.guard.ts [hunk 41]
```diff
@@ -15,7 +15,7 @@ export class AuthGuard implements CanActivate {
 	): Observable<boolean> | Promise<boolean> | boolean {
 		this.authService.isLoggedIn.subscribe((state) => {
 			if (!state) {
-				this.router.navigate([Routes.SIGN_IN]);
+				this.router.navigate(["home-page"]);
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/authorization/guards/auth/auth.guard.ts]: done

File: projects/core/src/lib/components/header/header.component.ts [hunk 42]
```diff
@@ -0,0 +1,32 @@
...
+	selector: "jsmu-header",
+	templateUrl: "./header.component.html",
+	styleUrls: ["./header.component.scss"],
+	changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class HeaderComponent implements OnInit {
+	public isLoggedIn!: Observable<User | null>;
+	public menuItems = menuItems;
+	public isSideBarOpen = false;
+	public spans = [1, 2, 3];
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/core/src/lib/components/header/header.component.ts]: done

File: projects/common-components/src/lib/navigation/navigation-item-list/navigation-item-list.component.html [hunk 43]
```diff
@@ -0,0 +1,6 @@
+<ul class="jsmu-nav-item-list">
+	<jsmu-navigation-item
+		*ngFor="let item of navigationItems"
+		[itemObj]="item"
```
[the remark this answers is not in the document]
Comment (KenanHaciyev) [reply projects/common-components/src/lib/navigation/navigation-item-list/navigation-item-list.component.html]: done

**Pull request merged**